sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/VarShouldBeValCheck.kt

```diff
@@ -0,0 +1,126 @@
+
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtBinaryExpression
+import org.jetbrains.kotlin.psi.KtCallableReferenceExpression
+import org.jetbrains.kotlin.psi.KtDestructuringDeclarationEntry
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.KtNamedDeclaration
+import org.jetbrains.kotlin.psi.KtParenthesizedExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtUnaryExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.isPrivate
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.BindingContext.DECLARATION_TO_DESCRIPTOR
+import org.jetbrains.kotlin.resolve.calls.util.getResolvedCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.checks.AbstractCheck
+import org.sonarsource.kotlin.api.frontend.KotlinFileContext
+
+@Rule(key = "S3353")
+class VarShouldBeValCheck : AbstractCheck() {
+
+    override fun visitKtFile(file: KtFile, data: KotlinFileContext) {
+        super.visitKtFile(file, data)
+        val bindingContext: BindingContext = data.bindingContext
+
+        val varProperties = file.collectDescendantsOfType<KtProperty> { it.localVar() || it.privateClassVar() }
+        val destructedVar = file.collectDescendantsOfType<KtDestructuringDeclarationEntry> { it.localVar() }
+        val allVars = varProperties + destructedVar
+
+        val binaryAssignments = file.collectDescendantsOfType<KtBinaryExpression> { it.isAssignment() }
+        val unaryAssignments = file.collectDescendantsOfType<KtUnaryExpression> { it.isAssignment() }
+        val callableVarReference = file.collectDescendantsOfType<KtCallableReferenceExpression>()
+
+        val assignedExpressions = listOf(binaryAssignments.mapNotNull { it.reference() },
+            unaryAssignments.mapNotNull { it.reference() },
+            callableVarReference.map { it.callableReference }).flatten()
+
+        val declarationToAssignment = assignedExpressions.groupBy { it.getResolvedCall(bindingContext)?.resultingDescriptor?.original }
+        val nameToAssignment = assignedExpressions.groupBy { it.getReferencedName() }
+
+        allVars.forEach { variable ->
+            val descriptor = bindingContext[DECLARATION_TO_DESCRIPTOR, variable]
+            val name = (variable as KtNamedDeclaration).name!!
+
+            val nameNotReferenced = nameToAssignment[name] == null
+            val isNotReferenced = declarationToAssignment[descriptor] == null
+
+            //first check in case we don't have semantics, second check use semantics
+            if (nameNotReferenced || (descriptor != null && isNotReferenced)) {
+                data.reportIssue(variable.valOrVarKeyword!!, """Replace the keyword `var` with `val`.""")
```

Comment:
It would be nice to have a very short reason in the message already. Something like for example "This property is never modified. Replace `var` with `val`"

Reply:
:+1: